Implement missing fields, methods, endpoints, and other values that are documented

vicky5124 opened the pull request:
Implements all the stuff that were documented on the discord API docs but not implemented in serenity.

The doc commits that are being implemented in this PR are:
https://github.com/discord/discord-api-docs/commit/53ee0109b874bde5ec2316379bf266bb332630a0
https://github.com/discord/discord-api-docs/commit/95816a58aec3b04798e77079dee2b271d243ecdd
https://github.com/discord/discord-api-docs/commit/30cc2ea5efc83e2398c72d2a47f5aeccd5e21a84
https://github.com/discord/discord-api-docs/commit/9114a1bcb882619f1c42dd1c577f22a80e82b125
https://github.com/discord/discord-api-docs/commit/7dbe9c760513daa2d3f88d52ba650a97ccb3daf9
https://github.com/discord/discord-api-docs/commit/d61166446ec954d2b8f561db0d87701df5de69d9
https://github.com/discord/discord-api-docs/commit/0f09124859874a48a44dadfea8197bfc7ff8b2fe
https://github.com/discord/discord-api-docs/commit/7ab5731e97262bcfdeb9278f8add0d555b98478d
https://github.com/discord/discord-api-docs/commit/ae09c938d05bd0ad2b5f2359ec3a08736f9a3b2c
https://github.com/discord/discord-api-docs/commit/aa5022364275072f775d162777cd5d3c348957ff
https://github.com/discord/discord-api-docs/commit/7684d7953bcdc50dd8c5ceffbd77c35b7d8a7fd5
https://github.com/discord/discord-api-docs/commit/52f1824451df49ad23bed19a706cbe124c7e7550
https://github.com/discord/discord-api-docs/commit/77ee0f906a3d9d2aff0bb3c285913e9db25d98dd

The commits that are not implemented in this PR and why are:
https://github.com/discord/discord-api-docs/commit/84f3290eb8155dac0a479e748f524542f2eda882 - Feature too big, should get its own PR
https://github.com/discord/discord-api-docs/commit/a80c210df2fba79ab95458d82f68633db7aaead6 - Requires previous point to be implemented.
https://github.com/discord/discord-api-docs/commit/ab9e3855ca2346075ee4d982a661cfa69a547260 - I have no idea how serenity handles this.


	modified:   src/builder/create_components.rs
	modified:   src/builder/create_embed.rs
	modified:   src/builder/create_message.rs
	modified:   src/builder/edit_message.rs
	modified:   src/builder/edit_role.rs
	modified:   src/builder/edit_thread.rs
	modified:   src/builder/execute_webhook.rs
	modified:   src/client/bridge/gateway/intents.rs
	modified:   src/client/dispatch.rs
	modified:   src/client/event_handler.rs
	modified:   src/http/client.rs
	modified:   src/http/routing.rs
	modified:   src/model/application.rs
	modified:   src/model/channel/attachment.rs
	modified:   src/model/channel/channel_id.rs
	modified:   src/model/channel/mod.rs
	modified:   src/model/event.rs
	modified:   src/model/gateway.rs
	modified:   src/model/guild/mod.rs
	modified:   src/model/guild/role.rs
	modified:   src/model/interactions/application_command.rs
	modified:   src/model/interactions/autocomplete.rs
	modified:   src/model/interactions/message_component.rs
	modified:   src/model/interactions/mod.rs
	modified:   src/model/interactions/ping.rs
	modified:   src/model/user.rs

vicky5124 committed:
Implement missing fields, methods, endpoints, and other values that are documented.
	modified:   src/builder/create_components.rs
	modified:   src/builder/create_embed.rs
	modified:   src/builder/create_message.rs
	modified:   src/builder/edit_message.rs
	modified:   src/builder/edit_role.rs
	modified:   src/builder/edit_thread.rs
	modified:   src/builder/execute_webhook.rs
	modified:   src/client/bridge/gateway/intents.rs
	modified:   src/client/dispatch.rs
	modified:   src/client/event_handler.rs
	modified:   src/http/client.rs
	modified:   src/http/routing.rs
	modified:   src/model/application.rs
	modified:   src/model/channel/attachment.rs
	modified:   src/model/channel/channel_id.rs
	modified:   src/model/channel/mod.rs
	modified:   src/model/event.rs
	modified:   src/model/gateway.rs
	modified:   src/model/guild/mod.rs
	modified:   src/model/guild/role.rs
	modified:   src/model/interactions/application_command.rs
	modified:   src/model/interactions/autocomplete.rs
	modified:   src/model/interactions/message_component.rs
	modified:   src/model/interactions/mod.rs
	modified:   src/model/interactions/ping.rs
	modified:   src/model/user.rs

vicky5124 committed:
fmt and clippy
	modified:   src/builder/create_message.rs
	modified:   src/builder/edit_message.rs
	modified:   src/builder/edit_role.rs
	modified:   src/builder/execute_webhook.rs
	modified:   src/client/event_handler.rs
	modified:   src/model/application.rs
	modified:   src/model/channel/channel_id.rs
	modified:   src/model/event.rs
	modified:   src/model/interactions/application_command.rs

vicky5124 commented on src/model/application.rs:
```diff
@@ -110,10 +115,7 @@ pub struct PartialCurrentApplicationInfo {
-    /// These flags are unknown and are not yet documented in the Discord API
-    /// documentation.
-    pub flags: u64,
+    pub flags: ApplicationFlags,
```
I think this is breaking

vicky5124 commented on src/model/application.rs:
```diff
@@ -26,7 +31,7 @@ pub struct ApplicationInfo {
     /// A set of bitflags assigned to the application, which represent gated
     /// feature flags that have been enabled for the application.
-    pub flags: Option<u64>,
+    pub flags: Option<ApplicationFlags>,
```
I think this is breaking

vicky5124 committed:
Fix breaking change.
	modified:   src/model/application.rs

vicky5124 committed:
Remove the breaking change.
	modified:   src/model/application.rs

vicky5124 committed:
fmt
	modified:   src/model/application.rs

vicky5124 committed:
Make `ApplicationFlags::from_u64` public
	modified:   src/model/application.rs

vicky5124 committed:
Fix tests
	modified:   src/model/interactions/mod.rs
	modified:   src/model/invite.rs
	modified:   src/model/misc.rs
	modified:   src/utils/mod.rs

arqunis commented on src/model/channel/attachment.rs:
```diff
@@ -30,6 +30,12 @@ pub struct Attachment {
+    /// Ephemeral attachments will automatically be removed after a set period of time.
+    /// Ephemeral attachments on messages are guaranteed to be available as long as
+    /// the message itself exists.
+    pub ephemeral: Option<bool>,
```
I think this should be a regular `bool` with the `#[serde(default)]` attribute. It is not clear based on the documentation about the state of the field, which has the potential to adopt one of the 3 states:
1. `None` -- unsure what this is supposed to represent, or when it may be realised; is it when the attachment isn't ephemeral, or...?
2. `Some(false)` -- the attachment is not ephemeral (or something else, 1. muddles this state)
3. `Some(true)` -- the attachment is ephemeral

```suggestion
    #[serde(default)]
    pub ephemeral: bool,
```

Furthermore, it might be wise to skip serialising the field if it is `false`:
```suggestion
    #[serde(default, skip_serializing_if = "is_false")]
    pub ephemeral: bool,
```
```rs
fn is_false(b: &bool) -> bool {
    !b
}
```

arqunis commented on src/model/channel/mod.rs:
```diff
@@ -479,6 +479,14 @@ pub struct ThreadMetadata {
+    /// Whether non-moderators can add other non-moderators to a thread.
+    ///
+    /// **Note**: Only available on private threads.
+    pub invitable: Option<bool>,
```
Ditto.

arqunis commented on src/model/interactions/mod.rs:
```diff
@@ -64,6 +64,16 @@ impl Interaction {
+    /// Gets the invoked guild locale.
+    pub fn guild_locale(&self) -> Option<&str> {
+        match self {
+            Interaction::Ping(ref i) => i.guild_locale.as_ref().map(String::as_str),
```
`ref` is implied when the expression in `match <expr>` is a reference.
```suggestion
            Interaction::Ping(i) => i.guild_locale.as_ref().map(String::as_str),
```

vicky5124 committed:
`ref` is implied when the expression in `match <expr>` is a reference.

Update src/model/interactions/mod.rs

Co-authored-by: Alex M. M. <acdenissk69@gmail.com>

vicky5124 committed:
Replace Optional fields with default values.
	modified:   src/internal/mod.rs
	modified:   src/internal/prelude.rs
	modified:   src/model/channel/attachment.rs
	modified:   src/model/channel/mod.rs

vicky5124 committed:
Merge remote-tracking branch 'refs/remotes/origin/missing_fields' into missing_fields

arqunis reviewed (approved):
Looks good, but before I'll merge, could you fix the `no default features` job? You just need to import `is_false` in `attachment.rs` explicitly, as the `internal::prelude` module is only imported there when the `model` feature is enabled.

vicky5124 committed:
Import is_false even if the model feature is not presesnt
	modified:   src/model/channel/attachment.rs
